df/df.1
.Dd December 2008
.Dt DF 1

.Sh NAME

.Nm df
.Nd report free disk space

.Sh SYNOPSIS

.Nm df
.Ar [-k][-P|-t][file...][Option End]

.Sh DESCRIPTION

    The df utility shall write the amount of available space [XSI] [Option
Start]  and file slots [Option End] for file systems on which the invoking
user has appropriate read access. File systems shall be specified by the
file operands; when none are specified, information shall be written for all
file systems. The format of the default output from df is unspecified, but
all space figures are reported in 512-byte units, unless the -k option is
specified. This output shall contain at least the file system names, amount
of available space on each of these file systems, [XSI] [Option Start]  and
the number of free file slots, or inodes, available; when -t is specified,
the output shall contain the total allocated space as well. [Option End]

.Sh OPTIONS

    The df utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -k
        Use 1024-byte units, instead of the default 512-byte units, when
writing space figures.
    -P
        Produce output in the format described in the STDOUT section.
    -t
        [XSI] [Option Start] Include total allocated-space figures in the
output. [Option End]

.Sh OPERANDS

    The following operand shall be supported:

    file
        A pathname of a file within the hierarchy of the desired file
system. If a file other than a FIFO, a regular file, a directory, [XSI]
[Option Start]  or a special file representing the device containing the
file system (for example, /dev/dsk/0s1) [Option End] is specified, the
results are unspecified. Otherwise, df shall write the amount of free space
in the file system containing the specified file operand.

.Sh STDIN

    Not used.

.Sh INPUT FILES

    None.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of df:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error and informative
messages written to standard output.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    When both the -k and -P options are specified, the following header line
shall be written (in the POSIX locale):

    "Filesystem 1024-blocks Used Available Capacity Mounted on\n"

    When the -P option is specified without the -k option, the following
header line shall be written (in the POSIX locale):

    "Filesystem 512-blocks Used Available Capacity Mounted on\n"

    The implementation may adjust the spacing of the header line and the
individual data lines so that the information is presented in orderly
columns.

    The remaining output with -P shall consist of one line of information
for each specified file system. These lines shall be formatted as follows:

    "%s %d %d %d %d%% %s\n", <file system name>, <total space>,
        <space used>, <space free>, <percentage used>,
        <file system root>

    In the following list, all quantities expressed in 512-byte units
(1024-byte when -k is specified) shall be rounded up to the next higher
unit. The fields are:

    <file system name>
        The name of the file system, in an implementation-defined format.
    <total space>
        The total size of the file system in 512-byte units. The exact
meaning of this figure is implementation-defined, but should include <space
used>, <space free>, plus any space reserved by the system not normally
available to a user.
    <space used>
        The total amount of space allocated to existing files in the file
system, in 512-byte units.
    <space free>
        The total amount of space available within the file system for the
creation of new files by unprivileged users, in 512-byte units. When this
figure is less than or equal to zero, it shall not be possible to create any
new files on the file system without first deleting others, unless the
process has appropriate privileges. The figure written may be less than
zero.
    <percentage used>
        The percentage of the normally available space that is currently
allocated to all files on the file system. This shall be calculated using
the fraction:

        <space used>/( <space used>+ <space free>)

        expressed as a percentage. This percentage may be greater than 100
if <space free> is less than zero. The percentage value shall be expressed
as a positive integer, with any fractional result causing it to be rounded
to the next highest integer.
    <file system root>
        The directory below which the file system hierarchy appears.

    [XSI] [Option Start] The output format is unspecified when -t is used.
[Option End]

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    On most systems, the "name of the file system, in an
implementation-defined format" is the special file on which the file system
is mounted.

    On large file systems, the calculation specified for percentage used can
create huge rounding errors.

.Sh EXAMPLES

        The following example writes portable information about the /usr
file system:

        df -P /usr

        Assuming that /usr/src is part of the /usr file system, the
following produces the same output as the previous example:

        df -P /usr/src

.Sh RATIONALE

    The behavior of df with the -P option is the default action of the 4.2
BSD df utility. The uppercase -P was selected to avoid collision with a
known industry extension using -p.

    Historical df implementations vary considerably in their default output.
It was therefore necessary to describe the default output in a loose manner
to accommodate all known historical implementations and to add a portable
option ( -P) to provide information in a portable format.

    The use of 512-byte units is historical practice and maintains
compatibility with ls and other utilities in this volume of IEEE Std
1003.1-2001. This does not mandate that the file system itself be based on
512-byte blocks. The -k option was added as a compromise measure. It was
agreed by the standard developers that 512 bytes was the best default unit
because of its complete historical consistency on System V (versus the mixed
512/1024-byte usage on BSD systems), and that a -k option to switch to
1024-byte units was a good compromise. Users who prefer the more logical
1024-byte quantity can easily alias df to df -k without breaking many
historical scripts relying on the 512-byte units.

    It was suggested that df and the various related utilities be modified
to access a BLOCKSIZE environment variable to achieve consistency and user
acceptance. Since this is not historical practice on any system, it is left
as a possible area for system extensions and will be re-evaluated in a
future version if it is widely implemented.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    find



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



df.c
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#define __STDC_FORMAT_MACROS 1

#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdbool.h>
#include <mntent.h>
#include <sys/vfs.h>
#include <paths.h>
#include <libpu.h>


static const char doc[] =
N_("df - report filesystem disk space usage");

#define args_doc file_args_doc

static struct argp_option options[] = {
	{ NULL, 'k', NULL, 0,
	  N_("Use 1024-byte units, instead of the default 512-byte units") },
	{ "portability", 'P', NULL, 0,
	  N_("Use POSIX-defined output format") },
	{ NULL, 't', NULL, 0,
	  N_("Include total allocated-space figures in the output.") },
	{ }
};

static int df_post_walk(struct walker *w);
static int df_pre_walk(struct walker *w);
static int df_arg(struct walker *w, const char *fn, const struct stat *lst);

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static struct walker walker;

struct fslist {
	char			*devname;
	char			*dir;
	dev_t			dev;
	bool			masked;
	struct fslist		*next;
};

static uintmax_t opt_block_size = 512;
static bool opt_portable;
static bool opt_total_alloc;
static bool fslist_masked;

static struct fslist *fslist, *fslist_last;


static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN

	case 'k':
		opt_block_size = 1024;
		break;

	PU_OPT_SET('P', portable)
	PU_OPT_SET('t', total_alloc)
	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

static void push_mntent(struct mntent *me)
{
	struct stat st;

	struct fslist *fl = (struct fslist *) xcalloc(1, sizeof(struct fslist));
	fl->devname = xstrdup(me->mnt_fsname);
	fl->dir = xstrdup(me->mnt_dir);

	if (stat(fl->devname, &st) == 0)
		fl->dev = st.st_rdev;
	else if (stat(fl->dir, &st) == 0)
		fl->dev = st.st_dev;
	else
		fl->dev = (dev_t) -1;

	if (!fslist) {
		fslist = fl;
		fslist_last = fl;
	} else {
		fslist_last->next = fl;
		fslist_last = fl;
	}
}

static int read_mntlist(void)
{
	struct mntent *me;

	FILE *f = setmntent(_PATH_MOUNTED, "r");
	if (!f) {
		perror(_PATH_MOUNTED);
		return 1;
	}

	while ((me = getmntent(f)) != NULL)
		push_mntent(me);

	endmntent(f);

	return 0;
}

static int df_arg(struct walker *w, const char *fn, const struct stat *st)
{
	struct fslist *tmp = fslist;

	fslist_masked = true;

	while (tmp) {
		if (tmp->dev == st->st_dev)
			tmp->masked = true;
		tmp = tmp->next;
	}

	return 0;
}

static void df_mask_all(void)
{
	struct fslist *tmp = fslist;
	while (tmp) {
		tmp->masked = true;
		tmp = tmp->next;
	}
}

static int df_output(struct fslist *fl)
{
	struct statfs sf;
	uintmax_t blksz, total, used, avail, free, pct;
	const char *fmt;

	if (statfs(fl->dir, &sf) < 0) {
		perror(fl->dir);
		return 1;
	}

	blksz = sf.f_bsize;

	total = sf.f_blocks;
	total *= blksz;
	total /= opt_block_size;

	avail = sf.f_bavail;
	avail *= blksz;
	avail /= opt_block_size;

	free = sf.f_bfree;
	free *= blksz;
	free /= opt_block_size;

	used = total - free;

	if (total == 0)
		return 0;

	pct = ((total - avail) * 100) / total;

	if (opt_portable)
		fmt = _("%-20s %9" PRIuMAX " %9" PRIuMAX " %9" PRIuMAX " %7" PRIuMAX "%% %s\n");
	else
		fmt = _("%-20s %9" PRIuMAX " %9" PRIuMAX " %9" PRIuMAX " %3" PRIuMAX "%% %s\n");

	printf(fmt,
	       fl->devname,
	       total,
	       used,
	       avail,
	       pct,
	       fl->dir);

	return 0;
}

static int df_output_fslist(void)
{
	struct fslist *tmp = fslist;
	int rc = 0;

	if (opt_portable)
		printf(_("Filesystem         %4ju-blocks      Used Available Capacity Mounted on\n"),
	       		opt_block_size);
	else
		printf(_("Filesystem         %4ju-blocks      Used Available Use%% Mounted on\n"),
	       		opt_block_size);

	while (tmp) {
		if (tmp->masked)
			rc |= df_output(tmp);
		tmp = tmp->next;
	}

	return rc;
}

static int df_pre_walk(struct walker *w)
{
	return read_mntlist();
}

static int df_post_walk(struct walker *w)
{
	if (!fslist_masked)
		df_mask_all();

	return df_output_fslist();
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.flags			= WF_STAT;
	walker.pre_walk			= df_pre_walk;
	walker.post_walk		= df_post_walk;
	walker.cmdline_arg		= df_arg;
	return walk(&walker, argc, argv);
}
